src/dataflow-types/src/proto/logging.rs

```diff
@@ -0,0 +1,240 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+include!(concat!(env!("OUT_DIR"), "/logging.rs"));
+
+use std::collections::HashMap;
+
+use crate::logging::{DifferentialLog, LogVariant, LoggingConfig, MaterializedLog, TimelyLog};
+use mz_expr::GlobalId;
+use mz_repr::proto::{ProtoRepr, TryFromProtoError, TryIntoIfSome};
+
+impl From<&LoggingConfig> for ProtoLoggingConfig {
+    fn from(x: &LoggingConfig) -> Self {
+        ProtoLoggingConfig {
+            granularity_ns: Some(x.granularity_ns.into_proto()),
+            active_logs: x.active_logs.iter().map(Into::into).collect(),
+            log_logging: x.log_logging,
+        }
+    }
+}
+
+impl TryFrom<ProtoLoggingConfig> for LoggingConfig {
+    type Error = TryFromProtoError;
+
+    fn try_from(x: ProtoLoggingConfig) -> Result<Self, Self::Error> {
+        Ok(LoggingConfig {
+            granularity_ns: x.granularity_ns.map(u128::from_proto).ok_or_else(|| {
```

Comment:
Hmmm, here you can *almost* use `try_into_if_some`, which has a blanket implementation, but for the wrong trait bound. Let me see if I can extend this so we can also cover types with a with `ProtoRepr` implementations.

Reply:
I think you can add this snippet right after the `impl<T, U> TryIntoIfSome<T> for Option<U>` definition.

```rust
/// Convenience syntax for trying to convert a `Self` value of type
/// `Option<U>` to `T` if the value is `Some(value)`, or returning
/// [`TryFromProtoError::MissingField`] if the value is `None`.
pub trait FromProtoIfSome<T> {
    fn from_proto_if_some<S: ToString>(self, field: S) -> Result<T, TryFromProtoError>;
}

/// A blanket implementation for `Option<U>` where `U` is the
/// `Proto$T` type for `T`.
impl<T> FromProtoIfSome<T> for Option<T::Repr>
where
    T: ProtoRepr,
{
    fn from_proto_if_some<S: ToString>(self, field: S) -> Result<T, TryFromProtoError> {
        T::from_proto(self.ok_or_else(|| TryFromProtoError::missing_field(field))?)
    }
}
```

Then you should be able to say

```rust
granularity_ns: x.granularity_ns.from_proto_if_some("ProtoLoggingConfig::granularity_ns")
```

note to self: Overall, I don't like how we integrate internal types and external types through different sets of traits. Ideally this should be done uniformly. For example, here a single blanket implementation for handling `Option<T>` if `T` is set should  be sufficient.